File: javascript/packages/linter/src/rules/html-head-only-elements.ts
import { HERB_ATTRIBUTES } from "@herb-tools/client/directives"

import { hasAttribute, getTagLocalName } from "@herb-tools/core"
import { isHeadOnlyTag, isBodyOnlyTag } from "../utils/rule-utils"

import { ParserRule } from "../types"
import { ElementStackVisitor } from "../utils/rule-utils"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types"
import type { ParserOptions, ParseResult, HTMLElementNode, ERBIfNode, ERBUnlessNode, ERBCaseNode, ERBCaseMatchNode } from "@herb-tools/core"

const inTheBodyNotTheHead = (ancestors: string[]) => ancestors.includes("body") && !ancestors.includes("head")

interface UndecidedElement {
  tagName: string
  node: HTMLElementNode
}

class HeadOnlyElementsVisitor extends ElementStackVisitor {
  private undecided: UndecidedElement[] = []
  private bodyOnlyTagName: string | null = null
  private conditionalDepth = 0

  visitERBIfNode(node: ERBIfNode): void {
    this.withinConditional(() => super.visitERBIfNode(node))
  }

  visitERBUnlessNode(node: ERBUnlessNode): void {
    this.withinConditional(() => super.visitERBUnlessNode(node))
  }

  visitERBCaseNode(node: ERBCaseNode): void {
    this.withinConditional(() => super.visitERBCaseNode(node))
  }

  visitERBCaseMatchNode(node: ERBCaseMatchNode): void {
    this.withinConditional(() => super.visitERBCaseMatchNode(node))
  }

  visitHTMLElementNode(node: HTMLElementNode): void {
    const tagName = getTagLocalName(node)

    if (tagName && isHeadOnlyTag(tagName)) {
      const isAllowedInSVG = (tagName === "title" || tagName === "style") && this.isInsideElement("svg")
      const isMetaWithItemprop = tagName === "meta" && hasAttribute(node, "itemprop")
      const isScopedStyle = tagName === "style" && (hasAttribute(node, "scoped") || hasAttribute(node, HERB_ATTRIBUTES.styleScoped))

      if (!isAllowedInSVG && !isMetaWithItemprop && !isScopedStyle) {
        const { verdict, chain } = this.placementAcrossCallers(inTheBodyNotTheHead)
        const message = `Element \`<${tagName}>\` must be placed inside the \`<head>\` tag.`

        if (verdict === "always") {
          this.addOffenseWithCallChain(message, node.open_tag?.location ?? node.location, chain)
        } else if (verdict === "mixed") {
          this.addOffenseWithCallChain(`${message} At least one call site renders this file inside the \`<body>\`.`, node.open_tag?.location ?? node.location, chain)
        } else if (verdict === "unknown" && this.alwaysRenders) {
          this.undecided.push({ tagName, node })
        }
      }
    } else if (tagName && isBodyOnlyTag(tagName) && this.alwaysRenders && !this.bodyOnlyTagName) {
      this.bodyOnlyTagName = tagName
    }

    super.visitHTMLElementNode(node)
  }

  reportContradictions(): void {
    if (!this.bodyOnlyTagName) return

    for (const { tagName, node } of this.undecided) {
      this.addOffense(
        `Element \`<${tagName}>\` must be placed inside the \`<head>\` tag. This template also renders the body-only element \`<${this.bodyOnlyTagName}>\`, so one of the two is misplaced.`,
        node.open_tag?.location ?? node.location,
      )
    }
  }

  private get alwaysRenders(): boolean {
    if (this.conditionalDepth > 0) return false
    if (this.isInsideDetachedBlock) return false
    if (this.isInsideElement("template")) return false

    return !this.isInsideElement("html", "head", "body")
  }

  private withinConditional(visit: () => void): void {
    this.conditionalDepth++
    visit()
    this.conditionalDepth--
  }
}

export class HTMLHeadOnlyElementsRule extends ParserRule {
  static autocorrectable = false
  static ruleName = "html-head-only-elements"
  static introducedIn = this.version("0.8.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error",
      exclude: ["**/*.xml", "**/*.xml.erb"],
      environments: ["cli", "browser"],
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      action_view_helpers: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new HeadOnlyElementsVisitor(this.ruleName, context)

    visitor.visit(result.value)
    visitor.reportContradictions()

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/html_head_only_elements.rs
use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::element_stack::ElementStack;
use crate::utils::html_data::HEAD_ONLY_TAG_NAMES;
use crate::utils::tag_utils::{get_open_tag, get_tag_name_from_element, has_attribute};

use herb::nodes::HTMLElementNode;
use herb::ParseResult;
use herb::Visitor;
use herb_config::Severity;

pub struct HTMLHeadOnlyElementsRule;

struct HeadOnlyElementsVisitor {
  rule_name: &'static str,
  offenses: Vec<UnboundOffense>,
  element_stack: ElementStack,
}

impl Visitor for HeadOnlyElementsVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    if let Some(tag_name) = get_tag_name_from_element(node) {
      let lowercase = tag_name.to_lowercase();

      if !self.element_stack.inside("head") && self.element_stack.inside("body") && HEAD_ONLY_TAG_NAMES.contains(lowercase.as_str()) {
        if (lowercase == "title" || lowercase == "style") && self.element_stack.inside("svg") {
          // allowed
        } else if lowercase == "meta" {
          if let Some(open_tag) = get_open_tag(node) {
            if !has_attribute(open_tag, "itemprop") {
              self.offenses.push(UnboundOffense::new(
                self.rule_name,
                format!("Element `<{}>` must be placed inside the `<head>` tag.", lowercase),
                node.location.clone(),
              ));
            }
          } else {
            self.offenses.push(UnboundOffense::new(
              self.rule_name,
              format!("Element `<{}>` must be placed inside the `<head>` tag.", lowercase),
              node.location.clone(),
            ));
          }
        } else {
          self.offenses.push(UnboundOffense::new(
            self.rule_name,
            format!("Element `<{}>` must be placed inside the `<head>` tag.", lowercase),
            node.location.clone(),
          ));
        }
      }

      self.element_stack.push(lowercase);
      self.walk_html_element_node(node);
      self.element_stack.pop();
    }
  }
}

impl Rule for HTMLHeadOnlyElementsRule {
  fn name(&self) -> &'static str {
    "html-head-only-elements"
  }

  fn default_severity(&self) -> Severity {
    Severity::Error
  }

  fn default_exclude(&self) -> &[&str] {
    &["**/*.xml", "**/*.xml.erb"]
  }
}

impl ParserRule for HTMLHeadOnlyElementsRule {
  fn check(&self, result: &ParseResult, _context: &LintContext) -> Vec<UnboundOffense> {
    let mut visitor = HeadOnlyElementsVisitor {
      rule_name: self.name(),
      offenses: Vec::new(),
      element_stack: ElementStack::new(),
    };

    visitor.visit_document_node(&result.value);

    visitor.offenses
  }
}
